genext2fs.c

```diff
@@ -2278,6 +2297,101 @@ add2fs_from_file(filesystem *fs, uint32 this_nod, FILE * fh, uint32 fs_timestamp
 		free(path2);
 }
 
+#ifdef HAVE_LIBARCHIVE
+// adds a tarball from stdin to the filesystem
+static void add2fs_from_tar(filesystem *fs, uint32 this_nod, int squash_uids, int squash_perms, uint32 fs_timestamp) {
+	int r;
+	uint32 nod, lnknod;
+	struct archive *a;
+	struct archive_entry *entry;
+	char *path2, *path3, *dir, *name, *lnk;
+	uint32 uid, gid, mode, ctime, mtime;
+	a = archive_read_new();
+	if (a == NULL)
+		error_msg_and_die("Couldn't create archive reader.");
+	if (archive_read_support_filter_all(a) != ARCHIVE_OK)
+		error_msg_and_die("Couldn't enable decompression");
+	if (archive_read_support_format_all(a) != ARCHIVE_OK)
+		error_msg_and_die("Couldn't enable read formats");
+
+	if ((r = archive_read_open_fd(a, 0, 10240)))
```

Comment:
Thank you for your input! Okay, so you'd rather add a new option like `-t` and the argument for that option would then be the tarball? Then using `-t -` could be made to mean "read from standard input" as the `-` is also used to mean "standard input" in other tools. Would that be okay?

Reply:
That would be okay, yes.

Reply:
Thinking a bit more about it, it would be high time to really separate `-d` and `-D`. So `dopt` shouldn't be a `char * dopt[MAX_DOPT]` but instead something like a `struct fspiece dopt[MAX_DOPT]` where
```c
struct fspiece
{
    char piece_type;
    char *piece_path;
}
```
Directories would have type `'d'`, device tables type `'D'`, tarballs type `'t'`. Or maybe an enum would be cleaner.
Would that be ok with you ?

Reply:
One more thing, I realized `-t` is already used by `mke2fs` with which I try to (very loosely) maintain options compatibility. So better use `-a`.

Reply:
I have done that part: there's now a `fslayer` struct (with already a FSLAYER_TAR type for you) which you should use for calling `add2fs_from_tar()` in `populate_fs()`. You should rebase.